[#5170] Add id to resource json serialization (master)

SwooshyCueb opened the pull request:
This is meant to address #5170. I'll make another pull request for the `4-2-stable` branch once this one has passed review.

If a test case is required, I may need some direction. I've tried adding one to `test_resource_administration.cpp`, but the resources added with `adm::client` do not show up in the output from `rcServerReport`.

trel commented:
code looks good to me - yes, let's discuss some tests...

and we're avoiding using the # autolinker until things are nice and tight - because it puts a notification on the issue itself, which can become noisy if done more than once/twice.

SwooshyCueb commented:
I've added a test to the python test suite that checks the output of `izonereport`. The test assumes that there is at least one zone actually present in the output, but I think that's an okay assumption to make.

alanking commented on scripts/irods/test/test_izonereport.py:
```diff
@@ -0,0 +1,32 @@
+            zone_report = json.loads(out)
+            self.assertGreaterEqual(len(out_j["zones"][0]["icat_server"]["resources"]), 1)
+            for resource in out_j["zones"][0]["icat_server"]["resources"]:
+                self.assertTrue('id' in resource)
```
Would it be worth verifying that the IDs are correct? I suppose that is more testing the resource property getter/setter than `izonereport`. Just a thought.

Also, minor: Are we going to be pulling `test_izonereport_with_coordinating_resources__ticket_3303` and `test_izonereport_key_sanitization` into this new suite? I recall talking about that, but maybe I am mistaken.

trel replied:
yes, we did.    let's get the izonereport tests into a single well-named spot.

SwooshyCueb replied:
Will do.

In regards to testing the validity of the ids, the c++ test I had written earlier covers this. I recall someone mentioning that the reason I was running into trouble was because I needed to use a new connection to call `rcServerReport`, so I can fix that up and commit it.

alanking replied:
Right, like you say, the issue from the C++ tests was that you were re-using the same connection in the administration library calls (which is certainly the intuitive thing to do). Unfortunately, the resource manager is not updated in the connected agent which is why the new resources were not appearing. In the python suite, we are using iCommands which results in a new connection for each call, so it should work if a query or `ilsresc` with some parsing is used.

Hope that helps!

alanking commented on server/core/src/irods_report_plugins_in_json.cpp:
```diff
@@ -97,6 +97,13 @@ namespace irods {
+        if ( !ret.ok() ) {
+            return PASS(ret);
+        }
+        std::string id_str = std::to_string( id );
```
Could the `std::to_string` call occur on 156 at the time of insertion into the `_entry` map?

SwooshyCueb replied:
Sure thing

alanking reviewed (approved):
Looks good overall. A couple of minor comments. Also, the commit message for the test should read "izonereport"

korydraughn commented on unit_tests/src/test_resource_administration.cpp:
```diff
@@ -77,6 +80,33 @@ TEST_CASE("resource administration")
+        REQUIRE(info->type() == ufs_info.resource_type);
+
+        bytesBuf_t *jbuf;
+        REQUIRE(rcServerReport((rcComm_t*)conn, &jbuf) == 0);
+        std::string report_str((char*)jbuf->buf, jbuf->len);
```
try to avoid using C style casts as they can introduce bugs.

prefer one of the C++ casting operators.  in this particular case, `static_cast`.

korydraughn commented on unit_tests/src/test_resource_administration.cpp:
```diff
@@ -77,6 +80,33 @@ TEST_CASE("resource administration")
+        REQUIRE(rcServerReport((rcComm_t*)conn, &jbuf) == 0);
+        std::string report_str((char*)jbuf->buf, jbuf->len);
+
+        auto report = nlohmann::json::parse(report_str);
```
this can be marked as `const` since the code only reads from it.

korydraughn commented on unit_tests/src/test_resource_administration.cpp:
```diff
@@ -77,6 +80,33 @@ TEST_CASE("resource administration")
+        auto report = nlohmann::json::parse(report_str);
+        REQUIRE(report["resources"].size() >= 2);
+        std::string id_str;
+        for (auto &resource : report["resources"]) {
```
`resource` can be marked `const` as well.  this can help the compiler produce more optimized code.

korydraughn commented on unit_tests/src/test_resource_administration.cpp:
```diff
@@ -77,6 +80,33 @@ TEST_CASE("resource administration")
+        REQUIRE(report["resources"].size() >= 2);
+        std::string id_str;
+        for (auto &resource : report["resources"]) {
+            std::string name = resource["name"].get<std::string>();
```
`name` can be marked `const`.

korydraughn commented on unit_tests/src/test_resource_administration.cpp:
```diff
@@ -77,6 +80,33 @@ TEST_CASE("resource administration")
+        REQUIRE(ec.value() == 0);
+        REQUIRE(info->type() == ufs_info.resource_type);
+
+        bytesBuf_t *jbuf;
```
prefer initializing your variables (unless you can make certain guarantees).  right now, this could point to anything in memory and cause the program to crash.

you can always use curly braces to zero out primitive types and C structures.  for example:
```c++
bytesBuf_t* bbuf{};  // Equivalent to: bytesBuf_t* bbuf = nullptr;
```

korydraughn reviewed (commented):
Looks good to me.

However, the test in the resource administration file should probably be moved to its own file.  The tests in the resource admin file verify that the C++ resource administration library produces the correct results.

The resource id test should have a dedicated file for verifying the behavior of `rcZoneReport`.

trel commented on unit_tests/src/test_zone_report.cpp:
```diff
@@ -0,0 +1,62 @@
+    // see issue #5170
+    SECTION("verify resource id from json serialization")
+    {
+        // we need a fresh connection becuase the cache is probably out of date
```
becuase -> because

trel commented:
please mark comments as resolved as you handle them - this is looking good.

SwooshyCueb commented:
I realized that the resources created in the setup function for the python test class don't actually show up in the `resources` array in the server report.
I've addressed this by adding some resources that do, and checking the ones that show up in the `coordinating_resources` array as well.

Additionally, I've also changed the test class to use the *class* setup/teardown functions and use a single session object for everything. This improves performance of the test class a bit, though this particular test class doesn't benefit as much from this change as others would, as it's only got four test methods, one of which is always skipped.

Is there anything else that needs to be addressed before I push this through jenkins?

trel commented:
i think there may be some duplication of resources listed in a multi-server zone - but that could be a separate issue.

consider an [858] on the second commit - otherwise, i think we're good.

SwooshyCueb commented:
All tests passed in jenkins except the one that doesn't, so I've added the #s

trel commented:
just caught a typo in the last two commit messages - 'resouce'->'resource'.

tweak and i'll merge.  thanks.

SwooshyCueb commented:
Whoops. Taken care of.

trel commented:
checkbox and cherry-pick for 4-2-stable por favor.
